### PR TITLE
fix some cases of `typeinfo` printing

### LINKED ISSUE #30683: 0/1 printing of false/true is confusing in Dict
```julia
julia> VERSION
v"1.2.0-DEV.146"

julia> show(Dict(true => false))
Dict(1=>0)
```

Also,

```julia
julia> Dict(true => false)
Dict{Bool,Bool} with 1 entry:
  true => false
```

seems to be inconsistent with

```julia
julia> [true => false]
1-element Array{Pair{Bool,Bool},1}:
 1 => 0
```

It looks like `display(Dict(...))` does not set `typeinfo` while `show(Dict(...))` does.  Here is the code I used to check it:

```julia
function logio_show(io, mime)
    push!(logio_log, ("mime", mime))
    push!(logio_log, ("io isa IOContext", io isa IOContext))
    if io isa IOContext
        push!(logio_log, ("io.dict", io.dict))
    end
    print(io, "LogIO")
end

Base.show(io::IO, mime::MIME"text/plain", ::LogIO) =
    logio_show(io, mime)

Base.show(io::IO, ::LogIO) =
    logio_show(io, nothing)
```

then

```julia
empty!(logio_log)
display(Dict(:a=>LogIO()))
logio_log
```

shows

```julia
3-element Array{Any,1}:
 ("mime", nothing)                                                                                                
 ("io isa IOContext", true)                                                                                       
 ("io.dict", Base.ImmutableDict{Symbol,Any}(:compact=>true,:SHOWN_SET=>Dict(:a=>LogIO),:module=>Main,:limit=>true,:color=>true))
```

(i.e., no typeinfo) while

```julia
empty!(logio_log)
show(Dict(:a=>LogIO()))
logio_log
```

shows

```
3-element Array{Any,1}:
 ("mime", nothing)                                                                                                
 ("io isa IOContext", true)                                                                                       
 ("io.dict", Base.ImmutableDict{Symbol,Any}(:typeinfo=>LogIO,:compact=>true,:compact=>true,:typeinfo=>Pair{Symbol,LogIO},:SHOWN_SET=>Dict(:a=>LogIO),:color=>true))
```

i.e., typeinfo is set to `LogIO`.

### PR DESCRIPTION
**Pull request by JeffBezanson**
Fixes #30683

Before:
```
julia> show(Dict(true=>false))
Dict(1 => 0)
```
after:
```
julia> show(Dict(true=>false))
Dict{Bool,Bool}(1 => 0)
```

Also fixes other similar cases with Dict, e.g.
```
julia> show(Dict(Any[1]=>Any[2]))
Dict([1] => [2])
```

As a bonus, there are also some reductions in verbosity. Certain types never need type info, e.g. `[1]` can always be shown instead of `Int[1]`. This adds `Symbol` and singleton types to that set.

### CODE REVIEW

**Comment by rfourquet** [comment]
This looks like an improvement in general, and the `isconcretetype` test was not a great heuristic, but there are cases where this change looks unnecessarily more verbose:
```julia
julia> show(Dict((1=>2) => (3=>4)))
Dict{Pair{Int64,Int64},Pair{Int64,Int64}}((1 => 2) => (3 => 4))
```
(Before: `Dict((1 => 2) => (3 => 4))`)

File: base/arrayshow.jl [hunk 1]
```diff
@@ -457,6 +457,11 @@ typeinfo_eltype(typeinfo::Type{<:AbstractArray{T}}) where {T} = eltype(typeinfo)
 typeinfo_eltype(typeinfo::Type{<:AbstractDict{K,V}}) where {K,V} = eltype(typeinfo)
 typeinfo_eltype(typeinfo::Type{<:AbstractSet{T}}) where {T} = eltype(typeinfo)
 
+# types that can be parsed back accurately from their un-decorated representations
+function typeinfo_implicit(@nospecialize(T))
+    return T === Float64 || T === Int || T === Char || T === String || T === Symbol ||
```
Comment (rfourquet) [inline base/arrayshow.jl]: Is it for compiler friendlyness that you don't write something like `T in (Float64, Int, Char, String, Symbol)` ?

**Comment by JeffBezanson** [comment]
> ```julia
> Dict{Pair{Int64,Int64},Pair{Int64,Int64}}((1 => 2) => (3 => 4))
> ```

I think we could make `typeinfo_implicit` recursive through Dict, Array, Pair, and Tuple element types.

File: test/core.jl [hunk 2]
```diff
@@ -1447,7 +1447,7 @@ struct Foo2509; foo::Int; end
 # issue #2517
 struct Foo2517; end
 @test repr(Foo2517()) == "$(curmod_prefix)Foo2517()"
-@test repr(Vector{Foo2517}(undef, 1)) == "$(curmod_prefix)Foo2517[$(curmod_prefix)Foo2517()]"
+@test repr(Vector{Foo2517}(undef, 1 )) == "[$(curmod_prefix)Foo2517()]"
```
Comment (JeffBezanson) [inline test/core.jl]:
```suggestion
@test repr(Vector{Foo2517}(undef, 1)) == "[$(curmod_prefix)Foo2517()]"
```

**Comment by JeffBezanson** [comment]
Ok, now everything looks real nice. We can remove a bunch more redundant type info too.

**Review by rfourquet — APPROVED** [review]
That's great!

**Pull request merged**